platform/android/src/jni.cpp

```diff
@@ -373,6 +435,77 @@ std::pair<mbgl::AnnotationSegment, mbgl::ShapeAnnotation::Properties> annotation
     return std::make_pair(segment, shapeProperties);
 }
 
+/*
+ * Adapted from https://github.com/dropbox/djinni/blob/master/support-lib/jni/Marshal.hpp
+ */
+
+void jniExceptionCheck(JNIEnv * env) {
+    if (!env) {
+        mbgl::Log::Error(mbgl::Event::JNI, "Empty environment.");
+        abort();
+    }
+
+    const jthrowable e(env->ExceptionOccurred());
+    if (e) {
+        env->ExceptionClear();
+    }
+}
+
+static std::vector<uint8_t> metadata_from_java(JNIEnv* jniEnv, jbyteArray j) {
+    assert(j != nullptr);
+
+    std::vector<uint8_t> ret;
+    jsize length = jniEnv->GetArrayLength(j);
+    jniExceptionCheck(jniEnv);
+
+    if (!length) {
+        mbgl::Log::Error(mbgl::Event::JNI, "jbyteArray was empty.");
+        return ret;
+    }
+
+    {
+        auto deleter = [jniEnv, j] (void* c) {
+            if (c) {
+                jniEnv->ReleasePrimitiveArrayCritical(j, c, JNI_ABORT);
+            }
+        };
+
+        std::unique_ptr<uint8_t, decltype(deleter)> ptr(
+            reinterpret_cast<uint8_t*>(jniEnv->GetPrimitiveArrayCritical(j, nullptr)),
+            deleter
+        );
+
+        if (ptr) {
+            // Construct and then move-assign. This copies the elements only once,
+            // and avoids having to initialize before filling (as with resize())
+            ret = std::vector<uint8_t>{ptr.get(), ptr.get() + length};
+        } else {
+            // Something failed.
+            jniExceptionCheck(jniEnv);
+        }
+    }
+
+    mbgl::Log::Error(mbgl::Event::JNI, "jbyteArray conversion completed.");
+    return ret;
+}
+
+static jbyteArray metadata_from_native(JNIEnv* jniEnv, const std::vector<uint8_t>& c) {
+    assert(c.size() <= std::numeric_limits<jsize>::max());
+    jbyteArray j = jniEnv->NewByteArray(static_cast<jsize>(c.size()));
+    jniExceptionCheck(jniEnv);
+
+    // Using .data() on an empty vector is UB
```

Comment:
As before, this has been rewritten: https://github.com/mapbox/mapbox-gl-native/commit/7f7f492219c29e800d190d54aa656c984a917538. Comments much appreciated.